money-http-async-client/src/main/scala/com/comcast/money/http/client/TraceFriendlyHttpAsyncClient.scala

```diff
@@ -0,0 +1,262 @@
+      // Stop timing the execution of the request...
+      tracer.stopTimer(HttpAsyncTraceConfig.HttpResponseTimeTraceKey)
+
+      // Get the response code, will be 0 if response is null
+      val responseCode = getResponseCode(response)
+      tracer.record(HttpAsyncTraceConfig.HttpResponseCodeTraceKey, responseCode)
+    })
+
+    // Continue with the execution of the request...
+    f(tracingCallback)
+  }
+
+  def wrapExecute[T](
+    requestProducer: HttpAsyncRequestProducer,
+    responseConsumer: HttpAsyncResponseConsumer[T],
+    callback: FutureCallback[T],
+    tracer: Tracer
+  )(f: (HttpAsyncRequestProducer, HttpAsyncResponseConsumer[T], FutureCallback[T]) => Future[T]): Future[T] = {
+    // capture the current tracing state...
+    val state = State.capture()
+    val span = SpanLocal.current
+
+    // Wrap the producer interface to intercept the request...
+    val tracingRequestProducer = new TracingHttpAsyncRequestProducer(requestProducer, state, (httpRequest: HttpRequest) => {
+      // Put the X-MoneyTrace header in the request...
+      TraceFriendlyHttpAsyncSupport.addTraceHeader(Option(httpRequest), span)
+
+      // Start timing the execution of the request...
+      tracer.startTimer(HttpAsyncTraceConfig.HttpResponseTimeTraceKey)
+      httpRequest
+    })
+
+    // Wrap the consumer interface to intercept the response...
+    val tracingResponseConsumer = new TracingHttpAsyncResponseConsumer[T](responseConsumer, state, (httpResponse: Try[HttpResponse]) => {
+      // Stop timing the execution of the request...
+      tracer.stopTimer(HttpAsyncTraceConfig.HttpResponseTimeTraceKey)
+
+      // Get the response code, will be 0 if response is null
+      val responseCode = getResponseCode(httpResponse)
+      tracer.record(HttpAsyncTraceConfig.HttpResponseCodeTraceKey, responseCode)
+      httpResponse
+    })
+
+    // Wrap the callback interface to restore the tracing state on the callback thread...
+    val tracingCallback = new TracingFutureCallback[T](Option(callback), state, _ => {})
+
+    // Continue with the execution of the request...
+    f(tracingRequestProducer, tracingResponseConsumer, tracingCallback)
+  }
+
+  def addTraceHeader(httpRequest: Option[HttpRequest], currentSpan: Option[Span]) {
+    (httpRequest, currentSpan) match {
+      case (Some(request), Some(span)) => request.setHeader("X-MoneyTrace", Formatters.toHttpHeader(span.info.id))
+    }
+  }
+
+  def getResponseCode(response: Try[HttpResponse]): Int =
+    response map (_.getStatusLine) map (_.getStatusCode) getOrElse 0
+}
+
+/**
+ * Provides a thin wrapper around [[HttpAsyncClient]] to support automatically tracing
+ * requests and restoring the tracing state on the callback interfaces.
+ */
+class TraceFriendlyHttpAsyncClient(wrappee: HttpAsyncClient) extends HttpAsyncClient
+    with java.io.Closeable {
+
+  import com.comcast.money.http.client.TraceFriendlyHttpAsyncSupport._
+
+  val tracer = Money.Environment.tracer
```

Comment:
would it make sense to pass in the tracer in the constructor?

Reply:
I was following the convention already set by [`TraceFriendlyHttpClient`](https://github.com/Comcast/money/blob/master/money-http-client/src/main/scala/com/comcast/money/http/client/TraceFriendlyHttpClient.scala#L75).  I don't mind breaking from that convention if it makes more sense.  Perhaps overloaded constructors for both?  Might make mocking easier rather than having to extend from the class which is what the unit tests currently do.

Reply:
Sorry about what is likely some shaky precedent.  I appreciate it that you tried to maintain consistency.  But yea, would make things easier just passing it in.

Reply:
Let me know if you can make that adjustment or if you need me to pair with you on this one.

Reply:
Got it